### PR TITLE
[Merged by Bors] - feat: port Data.List.Join

### PR DESCRIPTION
**Pull request by zeramorphic**
- [x] depends on: #1380

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1380
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by zeramorphic** [comment]
Following [Mario's advice](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EList.2EBasic/near/319679791), I modified the proof of `eq_iff_join_eq` to avoid using `nthLe` at all. This duplicates a couple of lemmas, so they now have `nthLe` and `get` variants. The `nthLe` variants are marked as deprecated.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1380~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by zeramorphic** [comment]
I think I've put the align in the right place.

File: Mathlib/Data/List/Join.lean [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+theorem join_append (L₁ L₂ : List (List α)) : join (L₁ ++ L₂) = join L₁ ++ join L₂ := by
+  induction L₁
+  · rfl
+  · simp [*]
+#align list.join_append List.join_append
+
+theorem join_concat (L : List (List α)) (l : List α) : join (L.concat l) = join L ++ l := by simp
+#align list.join_concat List.join_concat
+
+-- Porting note: `List.isEmpty` is missing alignment.
```
Comment (jcommelin) [inline Mathlib/Data/List/Join.lean:48]:
```suggestion
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Comment by jcommelin** [comment]
bors merge

**Pull request closed without merging**